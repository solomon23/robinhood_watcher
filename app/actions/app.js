// @flow
import * as API from '../services/api'
import { createRequests, doApiCall, createRequestTypes } from './actionHelper'
import { createSettingsWindow } from '../services/windows'

export const START_REFRESH_TIMER = 'START_REFRESH_TIMER'
export const STOP_REFRESH_TIMER = 'STOP_REFRESH_TIMER'
export const APP_QUIT = 'APP_QUIT'

export const GET_ACCOUNT_NUMBER = createRequestTypes('GET_ACCOUNT_NUMBER')
export const getAccountNumberActions = createRequests(GET_ACCOUNT_NUMBER)

export const GET_PORTFOLIO = createRequestTypes('GET_PORTFOLIO')
export const getPortfolioActions = createRequests(GET_PORTFOLIO)

export const GET_ACCOUNT_POSITIONS = createRequestTypes('GET_ACCOUNT_POSITIONS')
export const getAccountPositionsActions = createRequests(GET_ACCOUNT_POSITIONS)

export const GET_WATCH_LIST = createRequestTypes('GET_WATCH_LIST')
export const getWatchListActions = createRequests(GET_WATCH_LIST)

export function startRefreshTimer() {
  return { type: START_REFRESH_TIMER }
}

export function stopRefreshTimer() {
  return { type: STOP_REFRESH_TIMER }
}

export function appQuit() {
  return { type: APP_QUIT }
}

export function getAccount() {
  return doApiCall({
    api: API.getAccountNumber,
    action: getAccountNumberActions,
    guardFunc: (state) => {
      const { user } = state
      return !user.accountNumber
    },
  })
}

export function getPortfolio() {
  return doApiCall({
    api: API.getPortfolio,
    action: getPortfolioActions,
  })
}

export function getAccountPositions() {
  return doApiCall({
    api: API.getAccountPositions,
    action: getAccountPositionsActions,
  })
}

export function getWatchList() {
  return doApiCall({
    api: API.getWatchlist,
    action: getWatchListActions,
  })
}

export function getAllData() {
  return async function getAllCallback(dispatch: Dispatch, getState: GetState) {
    const { user } = getState()

    // bail if the user isn't logged in
    if (!user.authenticated) {
      return
    }

    if (!user.accountNumber) {
      await getAccount()(dispatch, getState)
    }

    await getPortfolio()(dispatch, getState)
    await getAccountPositions()(dispatch, getState)
    await getWatchList()(dispatch, getState)
  }
}

export function openSettings(settings: UserSettings) {
  return () => {
    createSettingsWindow(settings)
  }
}
