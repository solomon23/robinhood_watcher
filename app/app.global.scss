/*
 * @NOTE: Prepend a `~` to css file paths that are in your node_modules
 *        See https://github.com/webpack-contrib/sass-loader#imports
 */
@import "~@fortawesome/fontawesome-free/css/all.css";

$green: #21ce99;
$red: #f45531;

body {
  position: relative;
  color: white;
  height: 100vh;
  background-color: #212025;
  font-family: Arial, Helvetica, Helvetica Neue, serif;
  overflow-y: hidden;
  margin: 0;
}
