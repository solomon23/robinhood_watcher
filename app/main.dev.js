/* eslint global-require: 0, flowtype-errors/show-errors: 0 */

/**
 * This module executes inside of electron's main process. You can start
 * electron renderer process from here and communicate with the other processes
 * through IPC.
 *
 * When running `yarn build` or `yarn build-main`, this file is compiled to
 * `./app/main.prod.js` using webpack. This gives us some performance wins.
 *
 */
import { app, ipcMain } from 'electron';
import path from 'path';
import Toolbar from './toolbar';

let mainWindow = null;

if (process.env.NODE_ENV === 'production') {
  const sourceMapSupport = require('source-map-support');
  sourceMapSupport.install();
}

if (
  process.env.NODE_ENV === 'development' ||
  process.env.DEBUG_PROD === 'true'
) {
  require('electron-debug')();
  const p = path.join(__dirname, '..', 'app', 'node_modules');
  require('module').globalPaths.push(p);
}

const installExtensions = async () => {
  const installer = require('electron-devtools-installer');
  const forceDownload = !!process.env.UPGRADE_EXTENSIONS;
  const extensions = ['REACT_DEVELOPER_TOOLS', 'REDUX_DEVTOOLS'];

  return Promise.all(
    extensions.map(name => installer.default(installer[name], forceDownload))
  ).catch(console.log);
};

/**
 * Add event listeners...
 */

app.on('window-all-closed', () => {
  // Respect the OSX convention of having the application in memory even
  // after all windows have been closed
  if (process.platform !== 'darwin') {
    app.quit();
  }
});

app.on('ready', async () => {
  if (
    process.env.NODE_ENV === 'development' ||
    process.env.DEBUG_PROD === 'true'
  ) {
    await installExtensions();
  }

  mainWindow = Toolbar.create().window;

  ipcMain.on('SET_MENU_TITLE', (event, arg) => {
    Toolbar.setTitle(arg);
  });

  // pass messages to the main window
  ipcMain.on('MAIN_WINDOW', (event, arg) => {
    mainWindow.webContents.send(arg.channel, arg.message);
  });
});
