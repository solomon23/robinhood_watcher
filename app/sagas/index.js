import toolbarSaga from './toolbar'
import appSaga from './app'

export default (sagas) => {
  sagas.run(toolbarSaga)
  sagas.run(appSaga)
}
