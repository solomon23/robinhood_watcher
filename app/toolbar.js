import path from 'path';
import menubar from 'menubar';
import { Tray, nativeImage } from 'electron';

const ICON_LOGO_PATH = path.join(__dirname, '../resources/icons/16x16.png');
const ICON_LOGO = nativeImage.createFromPath(ICON_LOGO_PATH);

let mb = null;
let tray = null;

export default {
  setTitle: title => {
    if (mb && mb.tray) {
      mb.tray.setTitle(title);
    }
  },

  create: () => {
    tray = new Tray(ICON_LOGO);
    mb = menubar({
      dir: __dirname,
      icon: ICON_LOGO_PATH,
      preloadWindow: true,
      index: `file://${__dirname}/app.html?__dirname=${__dirname}`,
      width: 250,
      height: 500,
      tray,
      resizable: false,
      webPreferences: {
        experimentalFeatures: true
      }
    });

    mb.tray.setTitle(`-`);
    mb.showWindow();

    return mb;
  }
};
