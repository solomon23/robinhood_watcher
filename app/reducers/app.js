// @flow
import * as appActions from '../actions/app'

const defaultState = {
  actions: {
    portfolioLoading: false,
    positionsLoading: false,
    watchlistLoading: false,
    refreshing: false,
  },
  portfolio: {
  },
  positions: [
  ],
  watchlist: [
  ],
}

export default function(state: any = defaultState, action: Action) {
  switch (action.type) {
    case appActions.START_REFRESH_TIMER: {
      const actions = { ...state.actions, refreshing: true }
      return { ...state, actions }
    }

    case appActions.STOP_REFRESH_TIMER: {
      const actions = { ...state.actions, refreshing: false }
      return { ...state, actions }
    }

    case appActions.GET_PORTFOLIO.REQUEST: {
      const actions = { ...state.actions, portfolioLoading: true }
      return { ...state, actions }
    }

    case appActions.GET_PORTFOLIO.FAILURE: {
      const actions = { ...state.actions, portfolioLoading: false }
      return { ...state, actions }
    }

    case appActions.GET_PORTFOLIO.SUCCESS: {
      const actions = { ...state.actions, portfolioLoading: false }
      return { ...state, actions, portfolio: action.response }
    }

    case appActions.GET_ACCOUNT_POSITIONS.REQUEST: {
      const actions = { ...state.actions, positionsLoading: true }
      return { ...state, actions }
    }

    case appActions.GET_ACCOUNT_POSITIONS.FAILURE: {
      const actions = { ...state.actions, positionsLoading: false }
      return { ...state, actions }
    }

    case appActions.GET_ACCOUNT_POSITIONS.SUCCESS: {
      const actions = { ...state.actions, positionsLoading: false }
      return { ...state, actions, positions: action.response }
    }

    case appActions.GET_WATCH_LIST.REQUEST: {
      const actions = { ...state.actions, watchlistLoading: true }
      return { ...state, actions }
    }

    case appActions.GET_WATCH_LIST.FAILURE: {
      const actions = { ...state.actions, watchlistLoading: false }
      return { ...state, actions }
    }

    case appActions.GET_WATCH_LIST.SUCCESS: {
      const actions = { ...state.actions, watchlistLoading: false }

      // filter out the ones already in our positions
      const lookup = state.positions.reduce((p, c) => (
        { ...p, [c.symbol]: true }
      ), {})

      const watchlist = action.response.filter(w => !lookup[w.symbol])

      return { ...state, actions, watchlist }
    }

    default:
      return state
  }
}
