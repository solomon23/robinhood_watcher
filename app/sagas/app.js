import { takeEvery, fork, put, call, race, take } from 'redux-saga/effects'
import { ipcRenderer } from 'electron'
import * as appActions from '../actions/app'
import * as userActions from '../actions/user'
import * as api from '../services/api'

const REFRESH_TIME = 1000 * 60 * 10

const wait = ms => (
  new Promise(resolve => {
    setTimeout(() => resolve(), ms)
  })
)

function* handleStartRefresh() {
  yield takeEvery(appActions.START_REFRESH_TIMER, function* handler() {
    // fire it off first
    yield put(appActions.getAllData())

    while(true) {
      const timer = yield race({
        stopped: take(appActions.STOP_REFRESH_TIMER),
        tick: call(wait, REFRESH_TIME),
      })

      if (!timer.stopped) {
        yield put(appActions.getAllData())
      } else {
        break
      }
    }
  })
}

function* handleLogout() {
  yield takeEvery(userActions.USER_LOGOUT, function* handler() {
    // stop refreshing
    yield put (appActions.stopRefresh())

    // kill our cookie
    api.logout()
  })
}

function* appQuit() {
  yield takeEvery(appActions.APP_QUIT, () => {
    // send the quit
    ipcRenderer.send('APP_QUIT')
  })
}

function* app() {
  yield fork(handleStartRefresh)
  yield fork(handleLogout)
  yield fork(appQuit)
}

export default app
