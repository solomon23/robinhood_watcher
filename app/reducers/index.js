// @flow
import { combineReducers } from 'redux'
import { connectRouter } from 'connected-react-router'
import toolbar from './toolbar'
import user from './user'
import app from './app'
import settings from './settings'

export default function createRootReducer(history: {}) {
  const routerReducer = connectRouter(history)(() => {})

  return connectRouter(history)(
    combineReducers({
      router: routerReducer,
      toolbar,
      app,
      user,
      settings,
    })
  )
}
