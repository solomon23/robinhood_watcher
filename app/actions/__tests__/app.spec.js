import configureMockStore from 'redux-mock-store'
import fetchMock from 'fetch-mock'
import thunk from 'redux-thunk'
import * as actions from '../app'

const mockStore = configureMockStore([thunk])

describe('app actions', () => {
  afterEach(() => {
    fetchMock.restore()
  })

  it('should fire refresh timer', () => {
    expect(actions.startRefreshTimer()).toMatchSnapshot()
  })

  it('should fire stop timer', () => {
    expect(actions.stopRefresh()).toMatchSnapshot()
  })

  it('should fire getAccount', () => {
    fetchMock.mock('https://api.robinhood.com/accounts/',
      { body: { results: [{ account_number: 10 }] } }
    )
    const store = mockStore({ user: { authenticated: true, accountNumber: null } })

    return store.dispatch(actions.getAccount())
      .then(() => {
        expect(store.getActions()).toMatchSnapshot()
        return null
      })
  })

  it('should fire getPortfolio', () => {
    fetchMock.mock('https://api.robinhood.com/accounts/10/portfolio/',
      { body: { results: [{ portfolios: [] }] } }
    )
    const store = mockStore({ user: { authenticated: true, accountNumber: '10' } })

    return store.dispatch(actions.getPortfolio())
      .then(() => {
        expect(store.getActions()).toMatchSnapshot()
        return null
      })
  })

  it('should fire getAllData', () => {
    fetchMock.mock('https://api.robinhood.com/accounts/10/portfolio/',
      { body: { results: [] } }
    )
    fetchMock.mock('https://api.robinhood.com/accounts/10/positions/',
      { body: { results: [{ average_buy_price: 10, quantity: 1, instrument: 'https://instrument' }] } }
    )

    fetchMock.mock('https://instrument',
      { body: { quote: 'https://quote', symbol: 'a', name: 'name' } }
    )

    fetchMock.mock('https://quote',
      { body: { last_traded_price: 2 } }
    )

    const store = mockStore({ user: { authenticated: true, accountNumber: '10' } })

    return store.dispatch(actions.getAllData())
      .then(() => {
        expect(store.getActions()).toMatchSnapshot()
        return null
      })
  })
})
