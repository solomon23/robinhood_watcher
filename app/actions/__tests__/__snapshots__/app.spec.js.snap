// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app actions should fire getAccount 1`] = `
Array [
  Object {
    "request": Object {
      "user": Object {
        "accountNumber": null,
        "authenticated": true,
      },
    },
    "type": "GET_ACCOUNT_NUMBER_REQUEST",
  },
  Object {
    "request": Object {
      "user": Object {
        "accountNumber": null,
        "authenticated": true,
      },
    },
    "response": 10,
    "type": "GET_ACCOUNT_NUMBER_SUCCESS",
  },
]
`;

exports[`app actions should fire getAllData 1`] = `
Array [
  Object {
    "request": Object {
      "user": Object {
        "accountNumber": "10",
        "authenticated": true,
      },
    },
    "type": "GET_PORTFOLIO_REQUEST",
  },
  Object {
    "request": Object {
      "user": Object {
        "accountNumber": "10",
        "authenticated": true,
      },
    },
    "response": Object {
      "results": Array [],
    },
    "type": "GET_PORTFOLIO_SUCCESS",
  },
  Object {
    "request": Object {
      "user": Object {
        "accountNumber": "10",
        "authenticated": true,
      },
    },
    "type": "GET_ACCOUNT_POSITIONS_REQUEST",
  },
  Object {
    "request": Object {
      "user": Object {
        "accountNumber": "10",
        "authenticated": true,
      },
    },
    "response": Array [
      Object {
        "averageBuyPrice": 10,
        "currentPrice": 2,
        "instrument": Object {
          "name": "name",
          "quote": "https://quote",
          "symbol": "a",
        },
        "name": "name",
        "quantity": 1,
        "quote": Object {
          "last_traded_price": 2,
        },
        "symbol": "a",
      },
    ],
    "type": "GET_ACCOUNT_POSITIONS_SUCCESS",
  },
  Object {
    "request": Object {
      "user": Object {
        "accountNumber": "10",
        "authenticated": true,
      },
    },
    "type": "GET_WATCH_LIST_REQUEST",
  },
  Object {
    "request": Object {
      "user": Object {
        "accountNumber": "10",
        "authenticated": true,
      },
    },
    "response": Array [],
    "type": "GET_WATCH_LIST_SUCCESS",
  },
]
`;

exports[`app actions should fire getPortfolio 1`] = `
Array [
  Object {
    "request": Object {
      "user": Object {
        "accountNumber": "10",
        "authenticated": true,
      },
    },
    "type": "GET_PORTFOLIO_REQUEST",
  },
  Object {
    "request": Object {
      "user": Object {
        "accountNumber": "10",
        "authenticated": true,
      },
    },
    "response": Object {
      "results": Array [
        Object {
          "portfolios": Array [],
        },
      ],
    },
    "type": "GET_PORTFOLIO_SUCCESS",
  },
]
`;

exports[`app actions should fire getWatchList 1`] = `
Array [
  Object {
    "request": Object {
      "user": Object {
        "accountNumber": "10",
        "authenticated": true,
      },
    },
    "type": "GET_WATCH_LIST_REQUEST",
  },
  Object {
    "request": Object {
      "user": Object {
        "accountNumber": "10",
        "authenticated": true,
      },
    },
    "response": Array [
      Object {
        "last_traded_price": 2,
      },
    ],
    "type": "GET_WATCH_LIST_SUCCESS",
  },
]
`;

exports[`app actions should fire refresh timer 1`] = `
Object {
  "type": "START_REFRESH_TIMER",
}
`;

exports[`app actions should fire stop timer 1`] = `
Object {
  "type": "STOP_REFRESH_TIMER",
}
`;
